allennlp/tests/common/plugins_test.py

```diff
@@ -0,0 +1,190 @@
+import distutils.dir_util
+import tempfile
+
+import pytest
+from overrides import overrides
+
+from allennlp.commands import Subcommand
+from allennlp.common.plugins import (
+    discover_file_plugins,
+    discover_namespace_plugins,
+    discover_plugins,
+    import_plugins,
+)
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.common.testing.plugins_test_util import pip_install, push_python_project
+
+
+class TestPlugins(AllenNlpTestCase):
+    @overrides
+    def setUp(self):
+        super().setUp()
+        self.plugins_root = self.FIXTURES_ROOT / "plugins"
+        # "a" sets a "global" namespace plugin, because it's gonna be installed with pip.
+        self.project_a_fixtures_root = self.plugins_root / "project_a"
+        # "b" sets a "local" namespace plugin, because it's supposed to be run from that directory.
+        self.project_b_fixtures_root = self.plugins_root / "project_b"
+        # "c" sets a "global" namespace plugin, because it's gonna be installed with pip.
+        self.project_c_fixtures_root = self.plugins_root / "project_c"
+        # "d" sets a "local" file plugin, because it's supposed to be run from that directory
+        # and has a ".allennlp_plugins" file in it.
+        self.project_d_fixtures_root = self.plugins_root / "project_d"
+
+    def test_no_plugins(self):
+        available_plugins = set(discover_plugins())
+        self.assertSetEqual(set(), available_plugins)
+
+    def test_namespace_package_does_not_exist(self):
+        available_plugins = set(discover_namespace_plugins("dummy_namespace"))
+        self.assertSetEqual(set(), available_plugins)
+
+    def test_file_plugins_does_not_exist(self):
+        available_plugins = set(discover_file_plugins("dummy_file"))
+        self.assertSetEqual(set(), available_plugins)
+
+    def test_global_namespace_plugin(self):
+        with pip_install(self.project_a_fixtures_root, "a"):
+            available_plugins = set(discover_plugins())
+            self.assertSetEqual({"allennlp_plugins.a"}, available_plugins)
+
+            import_plugins()
+            subcommands_available = {t.__name__ for t in Subcommand.__subclasses__()}
+            self.assertIn("A", subcommands_available)
```

Comment:
Can you also test that A is not in the available subcommands before you do the pip install?  Same for all of the other tests.

Reply:
Hmm, I guess the plugin unloading bit makes me think that won't actually work?

Reply:
I think it should work because it's before, not after.

Reply:
Oh, you mean because it was installed in another test?

Reply:
Yeah.

Reply:
The issue is also that the subcommands keep registered because `Subcommand` is imported only once for the whole test suite. It's not easy to solve because it could potentially be implicitly and recursively imported even if you put an explicit import inside each test function.

So what I did is just to see if the plugins are discovered before each test function, which is a proxy of it.